fast-logger/System/Log/FastLogger/SingleLogger.hs

```diff
@@ -0,0 +1,94 @@
+{-# LANGUAGE RecordWildCards #-}
+
+module System.Log.FastLogger.SingleLogger (
+    SingleLogger(..)
+  , newSingleLogger
+  ) where
+
+import Control.Concurrent (forkIO, newEmptyMVar, MVar, takeMVar, putMVar)
+import Control.Concurrent.STM
+
+import System.Log.FastLogger.Imports
+import System.Log.FastLogger.LogStr
+import System.Log.FastLogger.Write
+
+----------------------------------------------------------------
+
+data SingleLogger = SingleLogger {
+    slgrRef  :: IORef (LogStr
+                      ,[LogStr] -- writer queue
+                      )
+  , slgrKill :: IO ()
+  , slgrWakeup :: IO ()
+  }
+
+instance Loggers SingleLogger where
+    stopLoggers = System.Log.FastLogger.SingleLogger.flushAllLog
+    pushAllLog  = System.Log.FastLogger.SingleLogger.pushAllLog
+    flushAllLog = System.Log.FastLogger.SingleLogger.shutdown
+
+----------------------------------------------------------------
+
+writer :: BufFD -> TVar Int -> IORef (LogStr, [LogStr]) -> MVar () -> IO ()
+writer buffd@BufFD{..} tvar ref mvar = loop (0 :: Int)
+  where
+    loop cnt = do
+        cnt' <- atomically $ do
```

Comment:
I'm wondering if you're already using `TVar` for spin-waiting, wouldn't it be easier to just use `TQueue` and a poison pill instead of `TVar`, `IORef` and additional `MVar` for signalling shutdown?

P.S. It took me a while to decipher the meaning of `tvar` , `ref`, `mvar`. Maybe it would be a good idea to use more meaningful names?

Reply:
STM is heavier than `atomicModifyIORef'`.
In this design, STM transactions run only when the builder reaches the limit.

Yes, `MVar` is not necessary. I should think whether or not we can remove it from `SingleLogger` easily.